### PR TITLE
Use event handler to capture messages in then_wait()

### PR DESCRIPTION
**Pull request by forslund**
## Description
A restructure on how waiting for events work in VK based on comments from @chrisveilleux in #2946.

This removes the handling of every single event by polling the new_message event, instead it sets up a temporary event handler handling the capture of events.

A check of the already captured messages are also made to avoid certain race conditions when messages are sent after the "when" but before the "then" is set up.

## How to test
Make sure the VK tests still passes.

## Contributor license agreement signed?
CLA [ Yes ]

### CODE REVIEW

**Comment by devops-mycroft** [comment]
Voight Kampff Integration Test Succeeded  ([Results](https://reports.mycroft.ai/core/PR-2951))

**Comment by forslund** [comment]
Current failure is due to some VK report rights issue, the actual tests passes

File: test/integrationtests/voight_kampff/tools.py [hunk 1]
```diff
@@ -23,8 +24,129 @@
...
+    def reset(self):
+        """Reset the wait state."""
+        self.result.clear()
+
+    # TODO: Remove in 21.08
+    def wait_unspecific(self, timeout):
+        """
+        Wait for a specified time for criteria to be fulfilled by any message.
+
+        This use case is deprecated and only for backward compatibility
```
Comment (krisgesling) [inline test/integrationtests/voight_kampff/tools.py:51]: Do you know where this is or was being used? Or just an edge case that might exist?

**Review by krisgesling — APPROVED** [review]
Much nicer! 

Yay for event driven processes, and the flow of `wait_specific` looks spot on to make sure things don't slip through the cracks.

Minor question that doesn't affect this being merged, just curious
